### ISSUE TITLE
Have a plan for planar format like YUV 420 and 422.

### ISSUE THREAD

**Issue by Kangz** [gpuweb/gpuweb#625]
It's often useful to process video frames using the GPU, and a lot of the times video frames come out of the encoder in the form of planar formats. Common ones (at least that I know about) are YUV 420 and 422.

We should have a plan for whether WebGPU should support planar formats, and which one it should support in core.

**Comment by shaoboyan** [comment]
(A quick thought)
Video is an important scenario on web and I fully agree  that WebGPU should try best to support it (e.g. handle NV12 or YUV422 format directly)
But I have some concerns if we choose to support concrete planar format directly in GPUTextureFormat.
The reason is that I think Js developers don't have abilities to get concrete planar format(e.g. NV12 or YUV422) from browser encoder/decoder (And I think the concrete planar format is decided by encoder/decoder implementation).  This may have potential compatible issue or implementation will take extra effort to do conversion. 
Is it possible to support planar formats by having a surface texture type which is backed by native planar format support abilities? (Also a surface sampler type too). This way will leave the concrete format details to implementation(dawn) and in there we should have more info.

**Comment by sandersdan** [comment]
We are implementing lower-level access to encoders and decoders (https://github.com/WICG/web-codecs), and exposing the concrete format is plausible (likely required for compute-like workloads over the output buffers).

For texturing, we can probably restrict to cases where the platform provides sampler extensions, and incur a copy otherwise.

Separately we also need to be able to import VideoFrames in some way. I see that buffers are Serializable, is there any discussion about that I can read?

**Comment by kainino0x** [comment]
If you're talking about GPUBuffers (and GPUTextures), they aren't traditionally serializable. They behave like SharedArrayBuffer: they have an internal object, and when you "serialize/deserialize" it to another thread you just get another ref to the same internal object. Mutable state is synchronized e.g. in the blink bindings code.

**Comment by Kangz** [comment]
We discussed this briefly with @kenrussell. The biggest design difficulty is that in native APIs, the planar format is baked in the pipeline object, that WebGPU doesn't want to dynamically recompile pipeline objects, and that the video is allowed to change planar formats in the middle. So it means that we can't create a single pipeline that will work for a whole video.

We found there at least the three way forward for the import path:

  1. Importing a YUV ImageBitmap in WebGPU does a conversion to RGB and the developer gets an RGB `GPUTexture`. (somewhat complex to implement, easy to use, so-so perf)

  2. WebCodec gives one ImageBitmap per plane and metadata for the developer, that has to deal with it themselves, by importing up to 3 IBs in `GPUTextures` and doing the sampling in the shader themselves. (easy to implement on the WebGPU side, hard to use for the developer, good perf).

 3. An idea originally by @jdashg for WebGL: produce a pessimizing pipelines that can take up to three textures and metadata to describe the conversions that need to happen, then do YUV to RGB conversion in software on the GPU, instead of hardcoding constants in the pipeline. The developer would declare a special "video texture" for the pipeline, and pass it the IB. (somewhat complex implementation on the WebGPU side, easy to use, good perf).

Option 3) seems like the most interesting from a developer point of view but needs careful investigation for feasibility.

One additional note is that if there is specialized YUV -> RGB conversion hardware in GPUs (that's not clear), then none of these options would be as efficient as using the native API's support for planar formats.

**Comment by sandersdan** [comment]
Note that it is plausible for the app developer to be responsible for detecting format changes and updating their pipeline when using WebCodecs, I wouldn't discount that for the most advanced users.

We'll plan for (1) for now then, since it is the same as the current WebGL path. (2) is a straightforward optimization for us after our MVP.

I'm not quite sure how (3) differs from (1), likely because I don't have a proper mental model for WebGPU.

It sounds like all of these options end up as 4:4:4. That's probably an indication that I should expect that for the foreseeable future and design WebCodecs accordingly.

**Comment by kainino0x** [comment]
(3) is essentially (very roughly) to provide:
- An opaque shader language object that encapsulates however many planes
are needed
- An opaque shader language method to pull an RGB pixel from that opaque
object (internally, we implement all of the necessary conversions; that
method dynamically, within each shader thread, chooses a codepath depending
on the decoded format).

**Comment by sandersdan** [comment]
Ah, basically a custom sampler. That would indeed be very cool. It would also be worth implementing where the samples are YUV (abstract just the data layout).

**Comment by kainino0x** [comment]
Yup, I was thinking about that too. It might even be reasonable to have only YUV samplers and not RGB, if we know there won't be any extra conversion magic we have to do (idk, color spaces?)

**Comment by sandersdan** [comment]
YUV to RGB is a linear operation, and we will be providing the matrix on VideoFrames.

Color spaces are a much deeper topic. I don't think we should be doing any conversions automatically; but typical implementations use a bilinear 3D LUT.

**Comment by kenrussell** [comment]
@sandersdan note that with the WEBGL_video_texture extension work ongoing - https://bugs.chromium.org/p/chromium/issues/detail?id=776222 - there's a hope to provide YUV textures to WebGL using a custom opaque sampler format. Perhaps we should discuss how this dovetails with the WebCodecs standardization process on https://github.com/KhronosGroup/WebGL .

**Comment by shaoboyan** [comment]
Yup. Basically, I think option 3) provides the similar concept like WEBGL_video_texture(opaque texture target and opaque sampler for this target) but can be more flexible.

**Comment by kodabb** [comment]
> YUV to RGB is a linear operation, and we will be providing the matrix on VideoFrames.

Sorry to butt in, but I think there should be a bit more clarification on those points: YUV to RGB is linear only in a very specific configuration, and that is usually not the case in video (aka the main use case for this feature).

In a more generic way, YUV to RGB conversion is subject to a non-linear transform conversion known as gamma (or inverse when going the other way), and the color space that can be represented in the YUV system need to be specified, usually as defined by a color space specification: a YUV buffer could be representing a tiny gamut (bt601) or a larger one (bt2020). Not only that but there could use cases where the user wants to perform tone mapping for custom applications, such as HDR: in that case the texture would also need export ancillary metadata (also known as the mastering display, and color light level).

In other words, only exporting the matrix is not enough, a YUV texture would need to have the full suite of properties: **matrix coefficients**, **color primaries** and **transform function** and **pixel range** used to represent the original YUV conversion.

> Color spaces are a much deeper topic. I don't think we should be doing any conversions automatically; but typical implementations use a bilinear 3D LUT.

I agree on this point, automatic color space conversion is a guaranteed way to miss a requirement and run into problems later. However at least in the video world bilinear 3D LUTs are not as common as GPUs, and what often happens is that transcoders take care of the conversion themselves, because the have the aforementioned proprieties available.

**Comment by sandersdan** [comment]
> In a more generic way, YUV to RGB conversion is subject to a non-linear transform conversion known as gamma

This is incorrect. YUV and RGB are linearly related, both describe points within a color space that has the properties you have described.

**Comment by kodabb** [comment]
> This is incorrect. YUV and RGB are linearly related, both describe points within a color space that has the properties you have described.

They are related sure, but not in a linear fashion: unless we're talking about different things, in a video file the YCbCr is always considered gamma-corrected (with the exception of when the transfer characteristic is set to linear) and as such you cannot use the YCbCr values as is, but you need to linearize it first.

Using your example, bt2020 specifies defines the primaries need for the color system conversion, but it assumes that the data is linear first, and as such it also defines a very important `Non-linear transfer function` -- this function describes how input data must be treated before conversion. The underlying specification for gamma functions, bt1886 describes in great length the functional operation that needs to happen for the signal to be usable, and it is an essential read before venturing any further.

On that note, please be careful with how you interpret a color specification, the classic BT.709, BT.601, BT.2020 specs are just "recommendations", in the sense that they recommend the video color properties to respect these parameters, but it's definitely possible that a video file contains different mix and matching functions from other specifications. Case in point, none of the BT specs (except BT.2100) onward define what happens to a full range signals, but software is allowed to support that.

**Comment by Kangz** [comment]
Either way, option (3) would still work if an additional 3D LUT needs to be provided or more conversion code written in the shader. As long as we now the set of possibilities before hand and can dynamically choose the correct possibility in the shader.

**Comment by haasn** [comment]
I wanted to offer my 2 cents. There are a number of issues with fixed YUV->RGB conversion, even if we allow the user to specify the matrix being used:

1. YUV data is typically subsampled, which would require access to the individual YUV planes in order to have the API be lossless. Having a magic RGBA sampler be the *only* way of accessing the output of decoder requires some form of interpolation be performed behind the scenes in the vast majority of cases, typically with little user control. (What about interpolation functions other than bilinear sampling? What about different chroma placements? What about obscure subsampling modes?)

2. YUV->RGB is only a linear transformation in a subset of cases. Examples to the contrary include BT.2020 constant luminance, BT.2100 ICtCp, or Dolby Vision Profile 5. Any API which only supports changing the matrix used will fail to support these (or future) inherently non-linear systems, forcing users to specify an identity matrix and do the conversions themselves.

3. YUV->RGB conversion involves a loss of precision if naively rounded. It's not uncommon to do such processing using a higher internal precision and dither before output. Any such RGBA-conversion sampler would therefore have to be sampled at a sufficient bit depth to avoid introducing distortion due to rounding.

4. In some contexts, users may want to perform additional post-processing on the raw YUV planes before further processing distorts them (e.g. debanding/deblocking, adding AV1 film grain, and more techniques that are needlessly difficult to do after reconstruction and RGB conversion).

Overall, regardless of any other "helpful" RGBA samplers being added to the APIs for the benefit of the less-informed users, my strong recommendation is to provide *some* way of accessing the raw planes as output by the (standards-conformant) decoders, without introducing additional (non-standardized) distortion on top due to forcing the use of your own reconstruction and color conversion routines. In simpler terms, I think that decoding APIs should be as transparent and lossless as possible; and forced YUV->RGB conversion is definitely not lossless and generally not transparent. And at the very least, if such a sampler is added, it **must** be possible for users to set the exact matrix used, because relying on the API implementers to keep up with changing standards and confusing metadata is a surefire way to reinvent the BT.601/BT.709 issue.

Similar issues have plagued e.g. DXVA, VDPAU and a whole heap of other video decoding APIs, where forced YUV->RGBA conversion as part of the decoder API ended up causing problems for video players like VLC and mpv; including quality degradation for subsampled content, quality degradation for high-bit-depth content, use of incorrect hard-coded matrices, and general headaches.

As for the original topic of the issue, I don't think planar GPU formats are required to enable planar access to YUV textures - exposing them as three (or two) separate textures seems fine to me. You say this approach is "hard for the developer", but as a developer of multimedia software, I think this approach makes it *easiest* for me.

**Comment by Kangz** [comment]
@sandersdan the comments above seem to be pointing more toward solution 2) where WebCodec gives metadata and a bunch of planes to the developer and they have to deal with them manually. There's little video-specific work to be done in WebGPU for this, it just needs to support importing / copying of some more ImageBitmap formats. Do you have a place in WebCodec where this issue can be discussed more at length?

**Comment by sandersdan** [comment]
The only issue we currently have related to this is https://github.com/WICG/web-codecs/issues/37 (planar access in WebGL).

I agree that an ImageBitmap per plane (R8/R16) is the obvious short-term path, the problem is that it's not symmetric; there is no reasonable way to produce a VideoFrame from a canvas context in this format to be sent to an encoder. If it's easy with WebGPU to export multiple buffers from a single computation, then we probably don't need to find another solution (instead we can focus on optimizing this path).

It's possible that developers would find a spatially-separated texture arrangement (eg. NV12) convenient enough that the current takeImageBitmap() path could handle subsampled YUV export, I don't have good signals on that at this time.

The current thinking is one ImageBitmap per frame for the MVP, and we'll pick a planar access path post-MVP.

**Comment by kainino0x** [comment]
I'm not sure whether ImageBitmap would be the right container object for those planes, especially given R8 and R16 are (I think) new backing formats for ImageBitmap that would:
(1) have to interoperate with all uses of ImageBitmap in the web platform
(2) have to be somehow guaranteed to map to the correct formats when used with WebGPU, even though the backing format of an ImageBitmap is (I think) supposed to be opaque

**Comment by sandersdan** [comment]
This is worth exploring. I imagine that the options are:

- Use a WebGPU type that supports moving/binding between contexts. (I'm assuming there is no plan for such a thing at this time.)
- Use another type that supports efficient use, like ImageBitmap. If ImageBitmap isn't generic enough, we have the options of improving it or making a new abstraction.
- Put our time into a binding extension, perhaps binding directly to a VideoFrame object. (If we can support RW bindings, then export is also solved.)

How do you think a better replacement for ImageBitmap would differ from ImageBitmap? It sounds like your concern is in needing to support conversions between formats if those are not made explicit?

**Comment by kenrussell** [comment]
ImageBitmap is supposed to represent an opaque, GPU-resident texture. Various interop paths have been added in the past for uploading depth camera textures from video elements to WebGL R16 textures - though those didn't touch ImageBitmap directly. I think on the whole that it would be best to explore using ImageBitmap to represent these planes, with additional metadata attached to them somehow. Once a prototype is working, the path toward standardization will be clearer.

**Comment by jdashg** [comment]
ImageBitmap is opaque so that it *may possibly* represent some non-CPU memory, but it's not guaranteed, and I recommend against treating/modeling ImageBitmap as a guaranteed-GPU-local resource. It's unfortunately more complicated than that.

**Comment by kenrussell** [comment]
Fair points, but ImageBitmap nonetheless represents an opaque handle for a graphics resource. Kai's points about making all of the interop paths work with these new-format ImageBitmaps are taken, but if a brand new opaque handle type is created for the WebCodec standard, then extensions or APIs have to be added to both WebGL and WebGPU specifically to upload or import them into sampleable textures. In my opinion, reusing the existing ImageBitmap type to represent the YUV planes will result in a cleaner overall standard.

**Comment by krogovin** [comment]
I can't stress how awful those copies can add up. Doing: readInto() --> ArrayBuffer then ArrayBuffer --> GPU means:

- The first operation is a copy from GPU to CPU and potentially with also an unswizzle to get it scan-line style
- The 2nd operation is a copy from CPU to GPU almost guaranteed to incur a swizzle reorder too.

That is a four buggers. For 4K with YUV420 that means 4 * 12,441,600 bytes per frame which is nearly 50MB/frame which at 60Hz gets to a lovely 3GB/s bandwidth used minimum. The swizzling and unswizzling can end up quite CPU expensive on those drivers/hardware where the swizzle/unwizzle is done by CPU on the driver; when that is the case, the CPU usage sky rockets.

That is going to be alot of overhead.

**Comment by sandersdan** [comment]
> readInto() --> ArrayBuffer then ArrayBuffer --> GPU means:

As currently implemented in Chrome, the readInto() operation is available only when the planar data is already in CPU memory. We don't expect to ever provide an implicit readback path, instead we are waiting for better primitives to expose the planar data we already have, the the format it is already in.

**Comment by krogovin** [comment]
Admittedly this is somewhat out of scope to ask for direct GPU surface from a VideoFrame, but the inability to directly use a decoded video frame's surfaces has massive negative performance impact. Indeed, the story of swizzle and unswizzle is that many drivers are not made for those paths to be fast because they don't expect every frame to be uploading or downloading a large block of texels.

I fervently hope that

> In the future, we'll be looking for a way to expose GPU planes in an efficient way.

comes true; also

> These are usually GPU buffers, but in some cases are single-channel textures.

I'd expect the Y-plane to be a a single channel texture and the UV to be a two channel texture; This is easily supportable by WebGPU since it already has one and two channel textures.

The awkward part will be if the textures are not 8-bits wide per channel.

**Comment by kvark** [comment]
Somebody should try writing down the statements and arguments into https://github.com/kvark/webgpu-debate/
There are too many factors in play, hard to keep it all in the mind cache.

**Comment by sandersdan** [comment]
> I'd expect the Y-plane to be a a single channel texture and the UV to be a two channel texture;

This would be an unusual arrangement. Typically VideoFrames are stored in GMB-equivalents, eg. IOSurface or DMABuf buffers. The most common format is NV12, where Y is alone and UV are interleaved. The most direct mapping for sampling from an NV12 buffer is single-channel, where the top two-thirds is Y samples, and the bottom portion alternates between U and V samples.

Most platforms go further and provide extensions to sample from these buffers as RGBA samples. On Android, such extensions are the standard way of accessing frames because the actual formats are esoteric and proprietary.

It is unusual for the native format to be actual textures, but such a thing could come out of a processing step (eg. from StretchRect() on Windows).

**Comment by haasn** [comment]
I fail to see the distinction between your description and the one you're
quoting.

Alternating samples *is* a "two-channel texture".

**Comment by krogovin** [comment]
> This would be an unusual arrangement. Typically VideoFrames are stored in GMB-equivalents, eg. IOSurface or DMABuf buffers. The most common format is NV12, where Y is alone and UV are interleaved. The most direct mapping for sampling from an NV12 buffer is single-channel, where the top two-thirds is Y samples, and the bottom portion alternates between U and V samples.

That is _**exactly**_  what I am saying. There are effectively two surfaces: a Y-surface and a UV-surface. The former is single channel and the latter is two channel. Indeed, this is also exactly what one does when getting the surfaces to be used by GL in libVA via EGL_EXT_image_dma_buf_import.

**Comment by sandersdan** [comment]
I am distinguishing between buffers (concrete layout in memory) and textures (concrete layout in space).

**Comment by krogovin** [comment]
Um, I think that distinguishing is .. not real. A YUV decoded frame has two surfaces: a Y surface and a UV surface. Assuming that the video decoder and 3D accelerator are reasonably well integrated, then those surfaces are useable by the samplers. One embodiment of that interface is EGL_EXT_image_dma_buf_import. In that case one creates an EGLImage for the Y-surface and an EGLImage for the UV surface and from those EGLImage's, one creates a two GL texture objects.

Underneath, nearly nothing is going on except that the meta-data (width, height, format, etc) for each of the surfaces is copied from the video API to the 3D API. The entire point is to avoid copying the actual data or for that matter performing format conversion.

On the open source side with Mesa on Intel GPU's with libVA powered by https://github.com/intel/media-driver , this works. As a side note, the people working on that video driver and Mesa have historically had nothing to do with each other.

**Comment by Kangz** [comment]
I'd like to close this now that WebGPU has the concept of `GPUExternalTexture` that can efficiently use YUV data. Any concerns @shaoboyan and others?

**Comment by shaoboyan** [comment]
No, I don't have requirements to support these special types. Agree to close it.
